File: backend/JDISCTF/schemas/challenge.py
"""Challenge marshmallow schemas"""
from flask_rebar import RequestSchema
from marshmallow import fields, Schema

from JDISCTF.schemas.category import CategorySchema
from JDISCTF.schemas.submission import TeamSubmissionSchema


class UserChallengeSchema(Schema):
    """Response schema for getting a challenge"""
    id = fields.Integer(required=True)
    category_id = fields.Integer(required=True)
    name = fields.String(required=True)
    description = fields.String(required=True)
    points = fields.Integer(required=True)
    solves = fields.Nested(TeamSubmissionSchema, required=True, many=True)
    is_solved = fields.Boolean()


class ChallengeByCategorySchema(CategorySchema):
    """Response schema for getting challenges grouped by category"""
    challenges = fields.Nested(UserChallengeSchema, many=True)


class SubmitFlagSchema(RequestSchema):
    """Schema for submitting a flag for a given challenge"""
    flag = fields.String(required=True)


class SubmitFlagResponseSchema(Schema):
    """Schema for the response of a flag submission"""
    correct = fields.Boolean(required=True)

File: frontend/user/src/models/challenge.ts
import { Submission } from '@/models/submission'

/**
 * Challenge model
 */
export class Challenge {
  public id: number;
  public name: string;
  public description: string;
  public points: number;
  public files: string[];
  public links: string[];
  public tags: string[];
  public is_solved: boolean;
  public solves: Submission[];

  constructor() {
    this.id = -1;
    this.name = '';
    this.description = '';
    this.points = 0;
    this.files = [];
    this.links = [];
    this.tags = [];
    this.is_solved = false;
    this.solves = [];
  }
}
